src/realm/parser/driver.cpp

```diff
@@ -643,26 +643,38 @@ Query BetweenNode::visit(ParserDriver* drv)
     Query q(drv->m_base_table);
 
     auto tmp = prop->visit(drv);
+
     const ObjPropertyBase* obj_prop = dynamic_cast<const ObjPropertyBase*>(tmp.get());
-    if (obj_prop) {
-        if (tmp->get_type() == type_Int) {
-            auto min_val = min->visit(drv, type_Int);
-            auto max_val = max->visit(drv, type_Int);
+
+    auto create_between_query = [&min, &max, &drv, &obj_prop](ConstTableRef table, DataType type) {
+        Query q(table);
+        auto min_val = min->visit(drv, type);
+        auto max_val = max->visit(drv, type);
+
+        if (type == type_Int) {
             q.between(obj_prop->column_key(), min_val->get_mixed().get_int(), max_val->get_mixed().get_int());
-            return q;
         }
-        if (tmp->get_type() == type_Timestamp) {
-            auto min_val = min->visit(drv, type_Timestamp);
-            auto max_val = max->visit(drv, type_Timestamp);
+        else if (type == type_Timestamp) {
             q.between(obj_prop->column_key(), min_val->get_mixed().get_timestamp(),
                       max_val->get_mixed().get_timestamp());
-            return q;
+        }
+        return q;
+    };
+
+    if (obj_prop) {
+        const auto type = tmp->get_type();
+        if (type == type_Int || type == type_Timestamp) {
+            if (obj_prop->links_exist()) {
+                const auto& link_map = obj_prop->get_link_map();
```

Comment:
I am not sure if this is enough, and it is covering all the cases, but it seems like it is enough to pass all core tests, we basically need to use the target table if the query is run over links.  @ironage @jedelbo

Reply:
This approach doesn't work because you are passing the destination column and the destination table, without accounting for the links along the way. The specialized between node doesn't handle links, so we can just fall back to a normal relational comparison below. I have opened https://github.com/realm/realm-core/issues/7935 to do it this way.

Reply:
Ok great! Probably we should have a test in core that tracks this, because if something is not working, some test should tell us!!! But again, thanks for spotting this.

Reply:
There is a test in https://github.com/realm/realm-core/pull/7965 which will catch this.